### PR TITLE
[gentoo] preper for usrmerge

### PR DESCRIPTION
**Pull request by 257**
upstream has added merged-usr profile: https://bugs.gentoo.org/843818#c2
this makes that profile default for us. I expect a stage3 for merged-usr
sometime soon which should result in signicant code-drop at our end.

Signed-off-by: Paymon MARANDI <darwinskernel@gmail.com>

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
What's the status here?

**Comment by 257** [comment]
i will rework it this weekend.

**Comment by 257** [comment]
> What's the status here?

actually it passes the test. alert are outdated/fixed. let me know if you see otherwise.

**Comment by DaanDeMeyer** [comment]
Any idea when we'll be getting the merged-usr stage3?

**Comment by 257** [comment]
soon i hope. i'm hoping that gentoo-releng's uses `mkosi` for at least generating the initial stage3s.
pinging @floppym

**Comment by 257** [comment]
> Any idea when we'll be getting the merged-usr stage3?

_by the EOY_, supposidly.

**Comment by DaanDeMeyer** [comment]
There's still a CI failure in the gentoo gpt ext4 CI run, so I don't think we can merge this just yet

**Comment by 257** [comment]
i will have a look over the weekend. yes, please hold off of merging this.

**Comment by 257** [comment]
> There's still a CI failure in the gentoo gpt ext4 CI run, so I don't think we can merge this just yet

possibly fails on `sys-kernel/gentoo-kernel-bin`...

builds here, stage3 images upstream are *autobuilt*. i even thought of freezing the build that works (today it's `stage3-amd64-nomultilib-systemd-20220911T170535Z.tar.xz` and `=sys-kernel/gentoo-kernel-bin-5.15.69`) but it's pointless because we don't have a guarantee from upstream that image would even be there at the time of next `mkosi` release. 

my understanding is that there is an effort upstream to stanardise the way these stage3 release work. 

i understand this breaks `mkosi`'s promise of 100% reproducibility but as you you can options are limited.

what do you think?

**Comment by 257** [comment]
> There's still a CI failure in the gentoo gpt ext4 CI run, so I don't think we can merge this just yet

hmm, actually [image boots](https://github.com/systemd/mkosi/actions/runs/3070063742/jobs/5098459142#step:9:995) [we fail on ssh](https://github.com/systemd/mkosi/actions/runs/3070063742/jobs/5098459142#step:9:1027) odd!

**Comment by 257** [comment]
seems to be a bug with portage and binary packages, fixed upstream. how do i force the actions to run again without pushing to the branch?

**Comment by DaanDeMeyer** [comment]
I restarted the failed job

**Comment by DaanDeMeyer** [comment]
@257 Can you take another look, something is borked with ssh

**Comment by 257** [comment]
> @257 Can you take another look, something is borked with ssh

i did while back i couldn't find a way to run the test locally. any pointers?

**Comment by 257** [comment]
> @257 Can you take another look, something is borked with ssh

ssh is fixed but now squash tests fail! don't merge; i will have another look.

**Comment by 257** [comment]
> > @257 Can you take another look, something is borked with ssh
> 
> ssh is fixed but now squash tests fail! don't merge; i will have another look.



> > @257 Can you take another look, something is borked with ssh
> 
> ssh is fixed but now squash tests fail! don't merge; i will have another look.

seems to be related to https://github.com/systemd/systemd/commit/a8040b6d0ad43e5cb7168fd3b739c8c68aaa61b1
@keszybz could you have a look? it only happens on plain_squash, systemd-251 both on github and my test environment.

**Comment by 257** [comment]
https://lore.kernel.org/lkml/20191130181548.GA28459@gentoo-tp.home/

**Comment by behrmann** [comment]
@257 can you rebase against main? `gentoo.py` has moved to `distributions/gentoo.py`. The rest should be pretty much intact (except for the interface talking to the rest of mkosi having moved into a subclass of `DistributionInstaller`).

**Comment by 257** [comment]
[disabled plain_squashfs supported for the time being](https://github.com/systemd/mkosi/actions/runs/3575374955/jobs/6011854813#step:7:51)

File: mkosi/distributions/gentoo.py
Comment (behrmann) [inline mkosi/distributions/gentoo.py:444]: It looks like you're doing all the work in the `__init__` method of this class. At this point it doesn't need to be a class anymore. You could just move all of it into this method, no?

Comment (257) [reply mkosi/distributions/gentoo.py:444]: one at the time :) moving things around like that is yet another big change.

Comment (257) [reply mkosi/distributions/gentoo.py:450]: is this necessary, `fedora` seems to be doing the same thing, albeit not inside a class.

**Comment by DaanDeMeyer** [comment]
@257 Can you rebase? plain_squashfs just got removed by #1276 so we don't need to disable it anymore

**Comment by 257** [comment]
np, working on it

**Comment by 257** [comment]
@DaanDeMeyer which version of `systemd-repart` comes with `--defer-partitions`? i'm hitting this here:
```
systemd-repart: unrecognized option '--defer-partitions'
```
```
systemd 252 (252)
+PAM -AUDIT -SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS -FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY -P11KIT +QRENCODE +TPM2 -BZIP2 +LZ4 -XZ -ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
```

**Comment by behrmann** [comment]
@257 see #1290

**Comment by 257** [comment]
> np, working on it

ready with possible exception of ssh. please review.

**Comment by 257** [comment]
> ready with possible exception of ssh. please review.
ssh is also fixed. probably a side-effect of usrmerge.

**Comment by 257** [comment]
not sure why centos is failing

**Comment by DaanDeMeyer** [comment]
> not sure why centos is failing

Unrelated to this PR

File: mkosi/distributions/gentoo.py
Comment (DaanDeMeyer) [inline mkosi/distributions/gentoo.py:206]: We should drop this, we don't install any fs tools by default for other distros anymore either

File: mkosi/distributions/gentoo.py
Comment (DaanDeMeyer) [inline mkosi/distributions/gentoo.py:367]: This should be done with host systemctl instead of image systemctl, so use `run()` instead of `run_workspace_command()`

Comment (DaanDeMeyer) [reply mkosi/distributions/gentoo.py:367]: Also, why do this here and not in the shared code we have for this?

File: mkosi/distributions/gentoo.py
Comment (DaanDeMeyer) [inline mkosi/distributions/gentoo.py:403]: Why does this have to run inside the image? We're trying to move all the package manager stuff to run outside of the image instead of inside the image

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
Looks good, a few comments though

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py:206]: even `btrfs-progs` and `dosfstools`?

Comment (DaanDeMeyer) [reply mkosi/distributions/gentoo.py:206]: Yup, because with systemd-repart, we don't know what filesystems will be in the final image, so we don't add anything by default, and leave it up to the user.

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py:403]: we ran into number of problems using host's `portage` in the past. only initial `sync_portage_tree()` uses host's `portage` and i'm tempted to even remove that.

also we still do make some changes related to `systemd` and we also install kernel related packages for bootable cases, see `pkgs['boot']` especially under `update_stage3()`.
also any extra packages asked by user also needs this, see `merge_user_pkgs()`.

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py]: we do if output is subvolume! unless i'm misunderstood what that means.

Comment (257) [reply mkosi/distributions/gentoo.py]: removing `dostfstools`

Comment (257) [reply mkosi/distributions/gentoo.py]: removed `btrfs-progs` too but i would be interested to know more about the case  `output == subvolume`.

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py]: don't remember why. i had a bit in `mkosi/__init__.py` related to this which got removed after switching to usrmerge. removed it to see what broke in the past, although it seems to me that CI doesn't have a test for ssh?

File: mkosi/distributions/gentoo.py
Comment (DaanDeMeyer) [reply mkosi/distributions/gentoo.py]: In case of `output == subvolume`, root directory is indeed a subvolume but that doesn't mean that the user won't pack up the subvolume inside an ext4 filesystem or something else later, just because the root directory is a subvolume does not mean that the user will want to have btrfs-progs available inside the root directory

File: mkosi/distributions/gentoo.py
Comment (DaanDeMeyer) [reply mkosi/distributions/gentoo.py:389]: I would prefer if we use host's portage if possible. We do the same for all other distributions. But we can leave that for another PR since this one is already been open for a long time.

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py]: ok! but now i would like to meet that user, lol.
well actually i just don't know a use case like that but i get that we need to stay as generic as possible.

**Review by DaanDeMeyer — APPROVED** [review]

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py:389]: i started out using host's `portage` and its python interface in face, which turned out to be horrible. even @poettering mentioned that we probably do the same for `dnf`...

yeah changing to host's would be a big big change with lots of test. although with usermerge in place we're dropping lots of code and complexity. i will have a go at it once day job gets a bit lighter.

File: mkosi/distributions/gentoo.py
Comment (behrmann) [inline mkosi/distributions/gentoo.py:237]: The parentheses are not balanced (why escape the dashes anyway?) and the regex doesn't match the example above, since the initial `[0-9TZ]+/` is and the `.xz` at the end are missing, how about
```suggestion
            # e.g.: 20221127T170156Z/stage3-amd64-nomultilib-systemd-mergedusr-20221127T170156Z.tar.xz
            regexp = f"^[0-9]+T[0-9]+Z/stage3-{arch}-nomultilib-systemd-mergedusr-[0-9]+T[0-9]+Z\.tar\.xz"
```

Comment (257) [reply mkosi/distributions/gentoo.py:237]: do we need the `[0-9TZ]+/` at the beginning and the `.xz` at the end?

Comment (behrmann) [reply mkosi/distributions/gentoo.py]: You tell me :) My point is mainly that the example and the regexp should actually match and document what is expected or needed. Also, the groups for dashes and dots are still a bit weird and the square brackets don't match (the one after `mergedusr` never closes).

Comment (257) [reply mkosi/distributions/gentoo.py]: you're right. fixed with your suggestion except for `\.`, i get `invalid escape sequence` here for some reason. python version? editor stuff?

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py]: seems fine, closing this.

File: mkosi/distributions/gentoo.py
Comment (257) [reply mkosi/distributions/gentoo.py:385]: keeping this unresolved for future ref.

File: mkosi/distributions/gentoo.py
Comment (behrmann) [reply mkosi/distributions/gentoo.py]: That's with your editor then, I assume, I ran this directly in the REPL (3.9 and 3.10)

Comment (257) [reply mkosi/distributions/gentoo.py]: changed to `\.`

**Pull request merged**
